text/0249-v2-experiments.md

```diff
@@ -358,17 +370,288 @@ Disadvantages:
+
+Option 5 was rejected since it relies on the use of prerelease qualifiers in a
+way that does not adhere to [Sematic Versioning](https://semver.org/). According
+to the semver [specification](https://semver.org/#spec-item-11), a precedence
+between versions is calculated **only** if the **major**, **minor** and
+**patch** are equal. In the above example, the range `^1.60.0` is not satisfied
+by `1.61.0-experimental` since the minor part is not equal. This becomes even
+more problematic in npm-v7, which automatically installs `peerDependencies`.
+With npm-v7, executing `npm install` will throw an exception when a consumer
+application declares a dependency on `aws-cdk-lib` version
+`1.61.0-experimental`, tries to consumes a library which declares a
+`peerDependency` on version `^1.60.0`. exception:
+
+`npm install` error:
+
+```
+npm ERR! code ERESOLVE
+npm ERR! ERESOLVE unable to resolve dependency tree
+npm ERR!
+npm ERR! While resolving: my-cdk-application@1.0.0
+npm ERR! Found: aws-cdk-lib@2.3.1-experimental
+npm ERR! node_modules/aws-cdk-lib
+npm ERR!   aws-cdk-lib@"2.3.1-experimental" from the root project
+npm ERR!
+npm ERR! Could not resolve dependency:
+npm ERR! peer aws-cdk-li@"^2.3.0" from my-construct-lib@1.6.0
+npm ERR! node_modules/my-construct-lib
+npm ERR!   my-construct-lib@"1.6.0" from the root project
+npm ERR!
+npm ERR! Fix the upstream dependency conflict, or retry
+npm ERR! this command with --force, or --legacy-peer-deps
+npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
+npm ERR!
+```
+
+Users can work around this by executing `npm install --force` but this is not a
+behavior we want to recommend.
+
+### Option 6: name experimental APIs with Experimental suffix/prefix
+
+The two major changes in v2, monolithic packaging, and no more breaking changes,
+share the same goal - encourage customers to declare a flexible version
+(`^`,`~`) on `aws-cdk-lib`. Fixed versions are a bad practice, they prevent
+users from getting updates that includes critical fixes, security patches and
+more. The mechanism we choose to implement experimental features should adhere
+to the same standard. This means that solutions such as releasing experimental
+features under a different module - `aws-cdk-lib-experimental`, or a different
+version line - `2.x.x-experimental` are not viable solutions. They might prevent
+users from declaring a fixed dependency on the `aws-cdk-lib` mainline release,
+but they mandate it in other release line we own and maintain.
+
+To allow the CDK team to keep experimenting while still adhering to
+[Semantic Versioning](https://semver.org/#semantic-versioning-200), we suggest
+experimental features will be a part of `aws-cdk-lib`, and maintain backward
+compatibility. Experimental features will be add to `aws-cdk-lib` using a
+specific naming scheme. For example, the first experimental version of the
+`grantWrite` API will be introduced under the name `grantWriteExp_v1`, when it's
```

Comment:
@RomainMuller how would the underscore work with jsii name translations?

Reply:
Today, only static readonly (constants) are allowed to have `_` in the name. [JSII8002](https://github.com/aws/jsii/blob/f3163e130b3907288584fbe31d1b43550876482f/packages/jsii/lib/jsii-diagnostic.ts#L644): "Method and property (unless they are static readonly) names must use camelCase"


I'll change it and remove the underscore